File: backend_py/store/mcp_servers.py
import json
from pathlib import Path
try:
    from config import DATA_DIR, get_mcp_servers_config
except ImportError:
    DATA_DIR = Path(__file__).resolve().parent.parent / "data"
    def get_mcp_servers_config():
        import json
        raw = __import__("os").environ.get("MCP_SERVERS_JSON", "").strip()
        if not raw:
            return []
        try:
            arr = json.loads(raw)
            return [s for s in arr if isinstance(s, dict) and isinstance(s.get("id"), str) and isinstance(s.get("command"), str)]
        except Exception:
            return []

MCP_SERVERS_FILE = DATA_DIR / "mcp-servers.json"

def _ensure_dir():
    MCP_SERVERS_FILE.parent.mkdir(parents=True, exist_ok=True)

def _read_all() -> dict:
    try:
        with open(MCP_SERVERS_FILE, "r", encoding="utf-8") as f:
            data = json.load(f)
            return data if isinstance(data, dict) else {}
    except (FileNotFoundError, json.JSONDecodeError):
        return {}

def _write_all(stored: dict):
    _ensure_dir()
    with open(MCP_SERVERS_FILE, "w", encoding="utf-8") as f:
        json.dump(stored, f, indent=2)

def _valid(s: dict) -> bool:
    return isinstance(s.get("id"), str) and isinstance(s.get("command"), str)

def get_user_mcp_servers(user_id: str) -> list[dict]:
    stored = _read_all()
    raw = stored.get(user_id)
    if not isinstance(raw, list):
        return []
    return [s for s in raw if _valid(s)]

def set_user_mcp_servers(user_id: str, servers: list[dict]) -> list[dict]:
    list_ = [s for s in servers if _valid(s)]
    stored = _read_all()
    stored[user_id] = list_
    _write_all(stored)
    return list_

def get_mcp_servers_config_for_user(user_id: str) -> list[dict]:
    from_store = get_user_mcp_servers(user_id)
    if from_store:
        return from_store
    return get_mcp_servers_config()

File: backend/src/store/mcp-servers.ts
/**
 * Per-user MCP server config stored in data/mcp-servers.json.
 * Same file-based pattern as connectors; no database required.
 */

import { readFile, writeFile, mkdir } from 'fs/promises';
import { join } from 'path';
import { config, getMcpServersConfig, type McpServerConfig } from '../config.js';

const MCP_SERVERS_FILE = join(config.dataDir, 'mcp-servers.json');

type Stored = Record<string, McpServerConfig[]>;

async function ensureDir(): Promise<void> {
  await mkdir(config.dataDir, { recursive: true });
}

async function readAll(): Promise<Stored> {
  try {
    const data = await readFile(MCP_SERVERS_FILE, 'utf-8');
    const parsed = JSON.parse(data);
    return typeof parsed === 'object' && parsed !== null ? parsed : {};
  } catch {
    return {};
  }
}

async function writeAll(stored: Stored): Promise<void> {
  await ensureDir();
  await writeFile(MCP_SERVERS_FILE, JSON.stringify(stored, null, 2), 'utf-8');
}

function isValidConfig(s: unknown): s is McpServerConfig {
  return (
    typeof s === 'object' &&
    s !== null &&
    typeof (s as McpServerConfig).id === 'string' &&
    typeof (s as McpServerConfig).command === 'string'
  );
}

export async function getUserMcpServers(userId: string): Promise<McpServerConfig[]> {
  const stored = await readAll();
  const raw = stored[userId];
  if (!Array.isArray(raw)) return [];
  return raw.filter(isValidConfig);
}

export async function setUserMcpServers(userId: string, servers: McpServerConfig[]): Promise<McpServerConfig[]> {
  const list = servers.filter(isValidConfig);
  const stored = await readAll();
  stored[userId] = list;
  await writeAll(stored);
  return list;
}

/** Config to use for a user: saved list if any, otherwise env MCP_SERVERS_JSON. */
export async function getMcpServersConfigForUser(userId: string): Promise<McpServerConfig[]> {
  const fromStore = await getUserMcpServers(userId);
  if (fromStore.length > 0) return fromStore;
  return getMcpServersConfig();
}
